### PR TITLE
implement 4 types of course list view

### PR DESCRIPTION
**Pull request by nalorakq8**
implement 4 types of course list view

### CODE REVIEW

File: courses/views.py [hunk 1]
```diff
@@ -339,18 +339,50 @@ def dispatch(self, *args, **kwargs):
...
-        request,
-        'course_list_to_add.html',
-        {
-            'courses': qs,
-            'student_id': student.pk
-        }
-    )
+    if not request.user.is_instructor():
+        qs = Course.objects.all()
+        student = get_object_or_404(Student, name=request.user)
```
Comment (almarzouq) [inline courses/views.py]: you don't have to fetch it this way

simply using:
`request.user.student`

will work

File: courses/templates/course_list_to_add.html [hunk 2]
```diff
@@ -12,12 +12,14 @@
     <th>Add</th>
   </tr>
   {% for c in courses %}
+  {% if c.student_registration_open %}
```
Comment (almarzouq) [inline courses/templates/course_list_to_add.html]: don't put this condition here,
put the condition in the query set filter
`Course.objects.filter(student_registration_open=True)`

**Comment by nalorakq8** [comment]
ok i will fix them

File: courses/templates/course_list_to_add.html (hunk 2 above)
Comment (almarzouq) [inline courses/templates/course_list_to_add.html]: in the templates .. use if statements to hide and show links .. not to do business logic

for example, link to join course, only show it for student
link to edit course or perform instructor actions, only show for instructor

File: courses/views.py (hunk 1 above)
Comment (almarzouq) [inline courses/views.py]: for instructor u can user
`request.user.instructor`

File: courses/views.py [hunk 3]
```diff
@@ -339,18 +339,50 @@ def dispatch(self, *args, **kwargs):
...
+def my_list_of_courses(request):
+    if not request.user.is_instructor():
+        qs = Course.objects.filter(students=request.user.student)
+        student = get_object_or_404(Student, name=request.user)
+        return render(
+            request,
+            'student_course_list.html',
+            {
+                'courses': qs,
+                'student_id': student.pk
```
Comment (almarzouq) [inline courses/views.py]: u don't need the student_id variable

replace it where it is used in the template with
`request.user.student.pk`

Comment (almarzouq) [inline courses/views.py]: also .. you can improve this function .. there is a lot of duplication in the code .. u can make it much better

after u receive student_id ... try to use render only once at the bottom .. and try to think how u might use less lines of code to do the same thing and in a more elegant manner

**Comment by almarzouq** [comment]
perfect .. but for the template name variable use `template` instead of `html` in the future

**Pull request merged**